File: nodejs/nodeI2C/i2c.c
#include "i2c.h"

int fd;

int _initI2C_(uint16_t addr) {
  fd = wiringPiI2CSetup(addr);

  return fd;
}

uint8_t readU8(uint16_t addr) {
  return (uint8_t)wiringPiI2CReadReg8(fd, addr);
}

int8_t readS8(uint16_t addr) {
  int8_t result = readU8(addr);
  if (result > 127)
    return result - 256;
  else
    return result;
}

uint16_t readU16(uint16_t addr) {
  uint8_t hi_byte = readU8(addr);
  uint16_t result = (hi_byte << 8) + readU8(addr + 1);
  return result;
}

int16_t readS16(uint16_t addr) {
  uint8_t hi_byte = readU8(addr);
  if (hi_byte > 127)
    hi_byte -= 256;
  int16_t result = (hi_byte << 8) + readU8(addr + 1);
  return result;
}

int write8(int addr, int data) {
  return wiringPiI2CWriteReg8(fd, addr, data);
}

File: python/I2C.py
#! /usr/bin/python

import smbus

class I2C :

    def __init__(self, address, bus = smbus.SMBus(1), debug = False) :
        self.address = address;
        self.bus = bus;
        self.debug = debug;

    def readU8(self, reg) : # read an unsigned byte value
        return self.bus.read_byte_data(self.address, reg);

    def readS8(self, reg) : # read a signed byte value
        result = self.readU8(reg);
        if (result > 127) :
            return result - 256;
        else :
            return result;
            
    def readU16(self, reg) : # read an unsigned 16-bit value
        hi_byte = self.readU8(reg);
        result = (hi_byte << 8) + self.readU8(reg + 1);
        return result;
    
    def readS16(self, reg) : # read a signed 16-bit value
        hi_byte = self.readU8(reg);
        if (hi_byte > 127) :
            hi_byte -= 256;
        result = (hi_byte << 8) + self.readU8(reg + 1);
        return result;

    def write8(self, cmd, val) :
        self.bus.write_byte_data(self.address, cmd, val);

    def writeList(self, reg, list) :
        self.bus.write_i2c_block_data(self.address, reg, list);
